vowpalwabbit/parse_regressor.cc

```diff
@@ -54,287 +49,307 @@ const size_t buf_size = 512;
+            "\n", 1, text);
+
+        // TODO: validate skips?
+        uint32_t skip_len = (uint32_t)all.skip_strings.size();
         text_len = sprintf_s(buff, buf_size, "%d skip: ", (int)skip_len);
-		bin_text_read_write_fixed(model_file, (char *)&skip_len, sizeof(skip_len),
-				"", read, 
-			buff, text_len, text);
+        bin_text_read_write_fixed_validate_eof(model_file, (char *)&skip_len, sizeof(skip_len),
+            "", read,
+            buff, text_len, text);
 
-      for (size_t i = 0; i < skip_len; i++)
-	{
-			char skip[4] = { 0, 0, 0, 0 };
-	  if (!read) {
+        for (size_t i = 0; i < skip_len; i++)
+        {
+            char skip[4] = { 0, 0, 0, 0 };
+            if (!read) {
                 text_len = sprintf_s(buff, buf_size, "%s ", all.skip_strings[i].c_str());
-	    memcpy(skip, all.skip_strings[i].c_str(), min(3, all.skip_strings[i].size()));
-	  }
-			bin_text_read_write_fixed(model_file, skip, 3,
-				    "", read,
-				buff, text_len, text);
-	  if (read)
-	    {
-				string temp(skip);
-	      all.skip_strings.push_back(temp);
-
-				all.args.push_back("--skips");
-				all.args.push_back(boost::lexical_cast<std::string>(temp));
-	    }
-	}
-		bin_text_read_write_fixed(model_file, buff, 0,
-				"", read, 
-			"\n", 1, text);
-      
+                memcpy(skip, all.skip_strings[i].c_str(), min(3, all.skip_strings[i].size()));
+            }
+
+            bin_text_read_write_fixed_validate_eof(model_file, skip, 3,
+                "", read,
+                buff, text_len, text);
+            if (read)
+            {
+                string temp(skip);
+                all.skip_strings.push_back(temp);
+
+                all.args.push_back("--skips");
+                all.args.push_back(boost::lexical_cast<std::string>(temp));
+            }
+        }
+        bin_text_read_write_fixed_validate_eof(model_file, buff, 0,
+            "", read,
+            "\n", 1, text);
+
         text_len = sprintf_s(buff, buf_size, "options:%s\n", all.file_options->str().c_str());
-		uint32_t len = (uint32_t)all.file_options->str().length() + 1;
+        uint32_t len = (uint32_t)all.file_options->str().length() + 1;
         memcpy(buff2, all.file_options->str().c_str(), min(len, buf_size));
         
-      if (read)
+        if (read)
         {
-	len = buf_size;
+            len = buf_size;
         }
         
-		bin_text_read_write(model_file, buff2, len,
-			  "", read,
-			  buff, text_len, text);
+        bin_text_read_write_validate_eof(model_file, buff2, len,
+            "", read,
+            buff, text_len, text);
         
-      if (read)
+        if (read)
         {
-	all.file_options->str(buff2);
-    }
+            all.file_options->str(buff2);
+        }
+
+        // Read/write checksum if required by version
+        if (all.model_file_ver >= VERSION_FILE_WITH_HEADER_HASH)
+        {
+            uint32_t check_sum = uniform_hash(model_file.space.begin, model_file.space.size(), 0);
```

Comment:
The problem starts here:

https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/io_buf.cc#L123

read() is not guaranteed to return the requested number of bytes. From 
the man page:

"It is not  an
        error  if  this  number  is smaller than the number of bytes 
requested;
        this may happen for example because fewer bytes are actually  
available
        right  now  (maybe  because we were close to end-of-file, or 
because we
        are reading from a pipe, or from a terminal),  or  because 
read()  was
        interrupted  by  a  signal."

Hence read_file:
https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/io_buf.h#L137 


may not fill the buffer.  Hence, if the number of calls to fill() differ 
from one run to another, you'll get different results.  In my 
experience, this only matters when reading from a pipe or socket, but 
given the unclear semantics...it feels iffy to trust that all OS/FS 
combinations read a complete blob when it happens to be available.

As far as what to do instead, I'd suggest adding a 'hash' flag to the 
io_buf datastructure, then lazily computing a chained hash when that 
flag is on.  This should allow interruptions at any point while 
maintaining consistency over a chain.

-John

On 09/20/2015 08:26 PM, Luong Hoang wrote:
>
> In vowpalwabbit/parse_regressor.cc 
> <https://github.com/JohnLangford/vowpal_wabbit/pull/757#discussion_r39937015>:
>
> >          {
> > -	all.file_options->str(buff2);
> > -    }
> > +            all.file_options->str(buff2);
> > +        }
> > +
> > +        // Read/write checksum if required by version
> > +        if (all.model_file_ver >= VERSION_FILE_WITH_HEADER_HASH)
> > +        {
> > +            uint32_t check_sum = uniform_hash(model_file.space.begin, model_file.space.size(), 0);
>
> Didn't realize the buffer depends on OS & FS. How about I keep a local 
> byte buffer of the header to compute the hash?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/JohnLangford/vowpal_wabbit/pull/757/files#r39937015>.
>